core/pkg/artifacts/file_cache.go

```diff
@@ -0,0 +1,163 @@
+package artifacts
+
+import (
+	"crypto/md5"
+	"encoding/base64"
+	"encoding/hex"
+	"errors"
+	"io"
+	"log/slog"
+	"os"
+	"path/filepath"
+
+	atomicfile "github.com/natefinch/atomic"
+	"github.com/wandb/wandb/core/pkg/utils"
+)
+
+type FileCache struct {
+	root string
+}
+
+func NewFileCache() *FileCache {
+	dir, found := os.LookupEnv("WANDB_CACHE_DIR")
+	if !found {
+		userCacheDir, err := os.UserCacheDir()
+		if err != nil {
+			return nil
+		}
+		dir = filepath.Join(userCacheDir, "wandb")
+	}
+	return &FileCache{root: filepath.Join(dir, "artifacts")}
+}
+
+// Copy a file into the cache and return the B64MD5 cache key.
+func (c *FileCache) AddFile(path string) (string, error) {
+	f, err := os.Open(path)
+	if err != nil {
+		return "", err
+	}
+	defer f.Close()
+	return c.Write(f)
+}
+
+// This is a convenience function intended for running in the background; instead of
+// returning an error it just logs it. This is often correct, since failing to add a
+// file to the cache shouldn't be a critical error.
+//
+// If a digest is passed in it will be checked against the calculated hash.
+func (c *FileCache) AddFileAndCheckDigest(path string, digest *string) {
+	b64md5, err := c.AddFile(path)
+	if err != nil {
+		slog.Error("Unable to add file to cache", "err", err)
+	}
+	if digest != nil && *digest != b64md5 {
+		slog.Error("File cache mismatch", "path", path, "expected", *digest, "actual", b64md5)
+	}
+}
+
+// Try to restore the file referenced in a manifest entry to the given destination.
+//
+// Returns true if the file is already there or is restored successfully.
+//
+// If the file exists, it will be hashed and overwritten if the hash is different.
+func (c *FileCache) RestoreTo(entry ManifestEntry, dst string) bool {
+	exists, _ := DeleteInvalid(dst, entry.Digest)
+	if exists {
+		return true
+	}
+	cachePath := c.pathTo(entry.Digest)
+	if cachePath == nil {
+		return false
+	}
+	// TODO (hugh): should we set the LocalPath in the entry to the dst?
+	return utils.CopyFile(*cachePath, dst) == nil
+}
+
+// Given the hash of a file return the base64 encoding and the internal cache path.
+func (c *FileCache) keyAndPath(hash []byte) (b64md5 string, cachePath string) {
+	b64md5 = base64.StdEncoding.EncodeToString(hash)
+	hexHash := hex.EncodeToString(hash)
+	return b64md5, filepath.Join(c.root, "obj", "md5", hexHash[:2], hexHash[2:])
+}
+
+func (c *FileCache) md5Path(b64md5 string) (string, error) {
+	hexHash, err := utils.B64ToHex(b64md5)
+	if err != nil {
+		return "", err
+	}
+	return filepath.Join(c.root, "obj", "md5", hexHash[:2], hexHash[2:]), nil
+}
+
+// If the file with the matching hash is in the cache, return the path to the file.
+func (c *FileCache) pathTo(b64md5 string) *string {
```

Comment:
best to avoid pointers to strings, it's too easy to get a nil dereference panic

this should either use the string zero value (empty string), or return a (string, error) pair